src/shared/utilities/timeoutUtils.ts

```diff
@@ -0,0 +1,50 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+/**
+ * Timeout coTimeoutbject that can handle both cancellation token- and time limit-style timeout situations.
+ * @param timeoutLength Length of timeout duration (in ms)
+ */
+export class Timeout {
+    private readonly _startTime: number
+    private readonly _endTime: number
+    private readonly _timer: Promise<boolean>
+    public constructor(timeoutLength: number) {
+        this._startTime = new Date().getTime()
+        this._endTime = this._startTime + timeoutLength
+        this._timer = new Promise<boolean>((resolve) => {
```

Comment:
rather than having this as an internal property of the class - what if this class extended `Promise` ? Effectively this `Timeout` is a promise with a time limit.

Reply:
Haven't thought about that, I'll give it a look.